pkg/operator/controller/instance_controller.go

```diff
@@ -0,0 +1,742 @@
+// +kubebuilder:rbac:groups="",resources=pods,verbs=list;get;watch
+// +kubebuilder:rbac:groups="apps",resources=deployments,verbs=list;get;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=clusterroles;clusterrolebindings,verbs=list;get;watch;create;update;patch;delete
+// TODO- should review the operator RABCs.
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *InstanceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&api.Instance{}).
+		Watches(
+			&appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: ControlPlaneName,
+				},
+			},
+			&handler.EnqueueRequestForObject{},
+		).
+		Watches(
+			&appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: DataPlaneName,
+				},
+			},
+			&handler.EnqueueRequestForObject{},
+		).
+		Complete(r)
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// The reconcile get instance YAML and creates the ClusterLink components.
+func (r *InstanceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// Fetch the Clusterlink instance
+	instance := &api.Instance{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		// Requeue the request on error
+		return ctrl.Result{}, err
+	}
+
+	// CRD details
+	r.Logger.Infof("Enter instance Reconcile - (Namespace: %s, Name: %s)", instance.Namespace, instance.Name)
+	r.Logger.Info("InstanceSpec- ",
+		" DataPlane.Type: ", instance.Spec.DataPlane.Type,
+		", DataPlane.Replicas: ", instance.Spec.DataPlane.Replicas,
+		", Ingress.Type: ", instance.Spec.Ingress.Type,
+		", LogLevel: ", instance.Spec.LogLevel,
+		", ContainerRegistry: ", instance.Spec.ContainerRegistry,
+		", ImageTag: ", instance.Spec.ImageTag,
+	)
+	r.instance = instance
+	if r.instance.Spec.ContainerRegistry != "" && r.instance.Spec.ContainerRegistry[len(r.instance.Spec.ContainerRegistry)-1:] != "/" {
+		r.instance.Spec.ContainerRegistry += "/"
+	}
+
+	// Set the deployment namespace
+	r.clNamespace = InstanceNamespace
+	if r.instance.Namespace != OperatorNameSpace { // Allow to create in any namespace. TODO -should remove this part and use only InstanceNamespace
+		r.clNamespace = r.instance.Namespace
+	}
+
+	// Examine DeletionTimestamp to determine if object is under deletion
+	if !r.instance.DeletionTimestamp.IsZero() {
+		if err := r.deleteAllResources(ctx); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if err := r.deleteFinalizer(ctx); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	// Set Finalizer if needed
+	if err := r.setFinalizer(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Set ClusterLink components if needed
+	if err := r.setClusterLink(ctx); err != nil {
```

Comment:
IMO `create`,  `generate` (or `apply` if you want to be k8s specific...) are better verbs than `set`.
nit 1: I think the comment is redundant - it does exactly what its name says.
nit 2: since this function (`setClusterLink`) is meant to do the opposite of `deleteAllResources`,  I would request that their names reflect that (e.g., create and delete ClusterLink as an example)